proxy/src/now.rs

```diff
@@ -0,0 +1,63 @@
+use std::time::Instant;
+
+pub trait Now: Clone {
```

Comment:
There are going to be things like @hawkw's recent work in Trust-DNS which aren't going to use this `Now` trait and instead are always going to use `Instant::now()`. @hawkw Do you think that you'd be able to write tests for Conduit's DNS functionality using this mechanism even if Trust-DNS doesn't use it?

* @hawkw Should Trust-DNS be using Tokio's `now()` instead of `Instant::now()`?

Reply:
@briansmith 
>  Do you think that you'd be able to write tests for Conduit's DNS functionality using this mechanism even if Trust-DNS doesn't use it?

I'm not sure. I think it may make at least some cases that we want to test easier, but it doesn't allow complete control of the time source without changes in Trust-DNS as well. 

Since Trust-DNS' API only exposes the deadlines (and not the TTL durations) of DNS lookups, and it uses `Instant::now()` to determine those deadlines, I think we could write tests for lookups timing out by mocking Conduit's timer and then advancing it past the deadlines of those lookups. I'm not sure if this would be sufficient to test _all_ the cases we want to cover.

> Should Trust-DNS be using Tokio's `now()` instead of `Instant::now()`?

This would certainly make testing easier, if both Conduit and Trust-DNS were using one time source that we have complete control over. But (as I explained above) it may not be strictly necessary...

Reply:
Yes, but there is _no_ solution that would satisfy this case without trust-dns being modified to expose a pluggable time source (which, imo, should probably be trust-dns specific and not a global tokio-level interface).

Reply:
If Trust-DNS used Tokio's `now()` and Tokio exposed its own mocking framework that let us control the value it returns from `now()` to Trust-DNS, then I think it would work.

Reply:
One of my core question is whether its within scope for tokio to provide a globally-useful (i.e. not timer-specific) abstraction for mocking times or whether tokio's scope ends at the timer use case. @carllerche

Reply:
Another option is changing Trust-DNS' API to expose TTLs either in addition to or instead of deadlines, so the API consumer has the option to choose the time source.

Reply:
It seems highly useful to have a common API for now() as well as a way to hook into that, so I would say it is definitely within the scope.